elements/lisk-tree/src/merkle_tree.ts

```diff
@@ -18,62 +18,155 @@ import { hash } from '@liskhq/lisk-cryptography';
 
+const LAYER_INDEX_SIZE = 1;
+const NODE_INDEX_SIZE = 8;
+const NODE_HASH_SIZE = 32;
+
+enum NodeType {
+	ROOT = 'root',
+	BRANCH = 'branch',
+	LEAF = 'leaf',
+}
+
+interface NodeData {
+	readonly value: Buffer;
+	readonly hash: Buffer;
+}
+
+interface NodeInfo {
+	readonly type: NodeType;
+	readonly hash: Buffer;
+	readonly value: Buffer;
+	readonly leftHash: Buffer;
+	readonly rightHash: Buffer;
+	readonly layerIndex: number;
+	readonly nodeIndex: bigint;
+}
+
+const isLeaf = (value: Buffer): boolean =>
+	value[0] === LEAF_PREFIX[0] && value.compare(Buffer.alloc(0)) !== 0;
+
 export class MerkleTree {
 	private _root: Buffer;
-	private _dataLength = 0;
-	private _data: { [key: string]: Buffer } = {};
+	private _width = 0;
+
+	// Object holds data in format { [hash]: value }
+	private _hashToBuffer: { [key: string]: Buffer } = {};
 
 	public constructor(initValues: Buffer[] = []) {
-		this._dataLength = initValues.length;
-		if (this._dataLength === 0) {
+		if (initValues.length === 0) {
 			this._root = EMPTY_HASH;
 			return;
 		}
-		const leafHashes = new Array<Buffer>(this._dataLength);
-		for (let i = 0; i < this._dataLength; i += 1) {
-			const value = Buffer.concat(
-				[LEAF_PREFIX, initValues[i]],
-				LEAF_PREFIX.length + initValues[i].length,
-			);
-			const key = hash(value);
-			leafHashes[i] = key;
-			this._data[key.toString('binary')] = value;
-		}
-		const height = this._getHeight();
-		let layer = leafHashes;
-		for (let i = 0; i < height - 1; i += 1) {
-			const currentLayerHashes: Buffer[] = [];
-			const pairBranches = layer.length >>> 1;
-			const remainder = layer.length & 1;
-			for (let j = 0; j < pairBranches; j += 1) {
-				const index = j * 2;
-				const value = Buffer.concat(
-					[BRANCH_PREFIX, layer[index], layer[index + 1]],
-					BRANCH_PREFIX.length + layer[index].length + layer[index + 1].length,
-				);
-				const key = hash(value);
-				this._data[key.toString('binary')] = value;
-				currentLayerHashes[j] = key;
-			}
-			if (remainder) {
-				const tmp = layer[layer.length - 1];
-				layer = currentLayerHashes;
-				layer.push(tmp);
-			} else {
-				layer = currentLayerHashes;
-			}
-		}
 
-		[this._root] = layer;
+		this._root = this._build(initValues);
 	}
 
 	public get root(): Buffer {
 		return this._root;
 	}
 
-	// eslint-disable-next-line
-	public append(_value: Buffer): { key: Buffer; value: Buffer }[] {
-		return [];
+	public getNode(nodeHash: Buffer): NodeInfo {
+		const value = this._hashToBuffer[nodeHash.toString('binary')];
+		// eslint-disable-next-line
+		if (!value) {
+			throw new Error(
+				`Hash does not exist in merkle tree: ${nodeHash.toString('binary')}`,
+			);
+		}
+
+		let type: NodeType;
+		if (this._root.compare(nodeHash) === 0) {
+			type = NodeType.ROOT;
+		} else if (isLeaf(value)) {
+			type = NodeType.LEAF;
+		} else {
+			type = NodeType.BRANCH;
+		}
+		const layerIndex = type === NodeType.LEAF ? 0 : value.readInt8(1);
+		const nodeIndex =
```

Comment:
I'm confused, doesn't this return 0 for leaf nodes? If that the case i think it is wrong.
As a general comment, why do we treat leaf and branches differently? (see also previous comment for layer index)

Reply:
You are correct it returns `0` for nodes. There is no way to determine nodeIndex from within this function with just the value Buffer, since it is not stored for leaf nodes. Therefore for now I opted to simply store 0 for now.

Reply:
Currently leaf node doesn't store nodeIndex, therefore you cannot determine it from within this function. The nodeIndex will properly be implemented in the next issue which tackles verification.